src/overlays/actors/ovl_Demo_Ext/z_demo_ext.c

```diff
@@ -1,17 +1,233 @@
+        previousCsAction = this->previousCsAction;
+
+        if (csAction != previousCsAction) {
+            switch (csAction) {
+                case 1:
+                    DemoExt_SetupWait(this);
+                    break;
+                case 2:
+                    DemoExt_SetupMaintainVortex(this, globalCtx);
+                    break;
+                case 3:
+                    DemoExt_SetupDispellVortex(this);
+                    break;
+                default:
+                    // Demo_Ext_Check_DemoMode: there is no such action!
+                    osSyncPrintf("Demo_Ext_Check_DemoMode:そんな動作は無い!!!!!!!!\n");
+                    break;
+            }
+            this->previousCsAction = csAction;
+        }
+    }
+}
+
+void DemoExt_SetScrollAndRotation(DemoExt* this) {
+    s16* scrollIncr = this->scrollIncr;
+    s16* curScroll = this->curScroll;
+    s32 i;
+
+    for (i = 3; i != 0; i--) {
+        curScroll[i] += scrollIncr[i];
+    }
+    this->rotationPitch += (s16)(kREG(34) + 1000);
+}
+
+void DemoExt_SetColorsAndScales(DemoExt* this) {
+    Vec3f* scales = &this->scales;
+    f32 shrinkFactor;
+
+    shrinkFactor = ((kREG(35) + 40.0f) - this->alphaTimer) / (kREG(35) + 40.0f);
+    if (shrinkFactor < 0.0f) {
+        shrinkFactor = 0.0f;
+    }
+
+    this->primAlpha = (u32)(kREG(28) + 0xFF) * shrinkFactor;
+    this->envAlpha = (u32)(kREG(32) + 0xFF) * shrinkFactor;
+    scales->x = (kREG(19) + 400.0f) * shrinkFactor;
+    scales->y = (kREG(20) + 100.0f) * shrinkFactor;
+    scales->z = (kREG(21) + 400.0f) * shrinkFactor;
+}
+
+void DemoExt_Wait(DemoExt* this, GlobalContext* globalCtx) {
+    DemoExt_CheckCsMode(this, globalCtx);
+}
+
+void DemoExt_MaintainVortex(DemoExt* this, GlobalContext* globalCtx) {
+    DemoExt_PlayVortexSFX(this);
+    DemoExt_SetScrollAndRotation(this);
+    DemoExt_CheckCsMode(this, globalCtx);
+}
+
+void DemoExt_DispellVortex(DemoExt* this, GlobalContext* globalCtx) {
+    DemoExt_PlayVortexSFX(this);
+    DemoExt_SetScrollAndRotation(this);
+    DemoExt_SetColorsAndScales(this);
+    DemoExt_FinishClosing(this);
+}
+
+void DemoExt_Update(Actor* thisx, GlobalContext* globalCtx) {
+    static DemoExtActionFunc sActionFuncs[] = { DemoExt_Wait, DemoExt_MaintainVortex, DemoExt_DispellVortex };
+    DemoExt* this = THIS;
+
+    if ((this->action < 0) || (this->action >= 3) || sActionFuncs[this->action] == NULL) {
+        // Main mode is abnormal!
+        osSyncPrintf(VT_FGCOL(RED) "メインモードがおかしい!!!!!!!!!!!!!!!!!!!!!!!!!\n" VT_RST);
+    } else {
+        sActionFuncs[this->action](this, globalCtx);
+    }
+}
+
+void DemoExt_DrawNothing(Actor* thisx, GlobalContext* globalCtx) {
+}
+
+void DemoExt_DrawVortex(Actor* thisx, GlobalContext* globalCtx) {
+    DemoExt* this = THIS;
+
+    Mtx* mtx;
+    GraphicsContext* gfxCtx;
+    s16* curScroll;
+    Vec3f* scales;
+
+    scales = &this->scales;
+    gfxCtx = globalCtx->state.gfxCtx;
+    mtx = Graph_Alloc(gfxCtx, 0x40);
+
```

Comment:
sizeof(Mtx)

Reply:
I know this one is easy, but is there a way to tell when to use sizeof in general?

Reply:
generally if an allocation is occurring, its probably not a hardcoded size (though it does happen). you may see a larger number that is really `sizeof(thing) * amount`